File: 2024/day20/java/Solution.java
import java.io.*;
import java.nio.file.*;
import java.util.*;

public class Solution {
    record Point(int r, int c) {}

    private char[][] grid;
    private int rows, cols;
    private Point start, end;

    public static void main(String[] args) throws IOException {
        var solution = new Solution();
        solution.run();
    }

    private void run() throws IOException {
        parseInput();
        var distances = tracePath();

        System.out.println("Part 1: " + countCheats(distances, 2, 100));
        System.out.println("Part 2: " + countCheats(distances, 20, 100));
    }

    private void parseInput() throws IOException {
        var lines = Files.readAllLines(Path.of("../input.txt"));

        rows = lines.size();
        cols = lines.getFirst().length();
        grid = new char[rows][cols];

        for (int r = 0; r < rows; r++) {
            var line = lines.get(r);
            for (int c = 0; c < cols; c++) {
                char ch = line.charAt(c);
                grid[r][c] = ch;
                if (ch == 'S') {
                    start = new Point(r, c);
                } else if (ch == 'E') {
                    end = new Point(r, c);
                }
            }
        }
    }

    private Map<Point, Integer> tracePath() {
        var distances = new HashMap<Point, Integer>();
        var queue = new ArrayDeque<Point>();

        distances.put(start, 0);
        queue.add(start);

        int[][] dirs = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

        while (!queue.isEmpty()) {
            var pos = queue.poll();

            if (pos.equals(end)) {
                break;
            }

            int currentDist = distances.get(pos);

            for (var dir : dirs) {
                int nr = pos.r() + dir[0];
                int nc = pos.c() + dir[1];

                if (nr >= 0 && nr < rows && nc >= 0 && nc < cols && grid[nr][nc] != '#') {
                    var next = new Point(nr, nc);
                    if (!distances.containsKey(next)) {
                        distances.put(next, currentDist + 1);
                        queue.add(next);
                    }
                }
            }
        }

        return distances;
    }

    private long countCheats(Map<Point, Integer> distances, int maxCheatTime, int minSavings) {
        var positions = new ArrayList<>(distances.entrySet());
        long count = 0;

        for (var entry1 : positions) {
            var p1 = entry1.getKey();
            int d1 = entry1.getValue();

            for (var entry2 : positions) {
                var p2 = entry2.getKey();
                int d2 = entry2.getValue();

                int cheatCost = Math.abs(p2.r() - p1.r()) + Math.abs(p2.c() - p1.c());

                if (cheatCost <= maxCheatTime) {
                    int savings = d2 - d1 - cheatCost;
                    if (savings >= minSavings) {
                        count++;
                    }
                }
            }
        }

        return count;
    }
}

File: 2024/day20/cpp/solution.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <queue>
#include <unordered_map>
#include <cstdlib>

struct Point {
    int r, c;

    bool operator==(const Point& other) const {
        return r == other.r && c == other.c;
    }
};

struct PointHash {
    size_t operator()(const Point& p) const {
        return std::hash<int>()(p.r) ^ (std::hash<int>()(p.c) << 16);
    }
};

using Grid = std::vector<std::string>;
using DistMap = std::unordered_map<Point, int, PointHash>;

struct ParseResult {
    Grid grid;
    Point start;
    Point end;
};

ParseResult parse_grid(const std::string& input) {
    ParseResult result;
    std::istringstream iss(input);
    std::string line;
    int r = 0;

    while (std::getline(iss, line)) {
        if (line.empty()) continue;
        result.grid.push_back(line);
        for (int c = 0; c < static_cast<int>(line.size()); ++c) {
            if (line[c] == 'S') {
                result.start = {r, c};
            } else if (line[c] == 'E') {
                result.end = {r, c};
            }
        }
        ++r;
    }

    return result;
}

DistMap trace_path(const Grid& grid, const Point& start, const Point& end) {
    int rows = static_cast<int>(grid.size());
    int cols = static_cast<int>(grid[0].size());

    DistMap dist;
    dist[start] = 0;

    std::queue<Point> queue;
    queue.push(start);

    const int dr[] = {-1, 1, 0, 0};
    const int dc[] = {0, 0, -1, 1};

    while (!queue.empty()) {
        Point curr = queue.front();
        queue.pop();

        if (curr == end) break;

        for (int i = 0; i < 4; ++i) {
            int nr = curr.r + dr[i];
            int nc = curr.c + dc[i];

            if (nr >= 0 && nr < rows && nc >= 0 && nc < cols && grid[nr][nc] != '#') {
                Point next = {nr, nc};
                if (dist.find(next) == dist.end()) {
                    dist[next] = dist[curr] + 1;
                    queue.push(next);
                }
            }
        }
    }

    return dist;
}

long long count_cheats(const DistMap& dist, int max_cheat_time, int min_savings) {
    long long count = 0;

    std::vector<std::pair<Point, int>> track_positions(dist.begin(), dist.end());

    for (const auto& [p1, d1] : track_positions) {
        for (const auto& [p2, d2] : track_positions) {
            int cheat_cost = std::abs(p2.r - p1.r) + std::abs(p2.c - p1.c);
            if (cheat_cost <= max_cheat_time) {
                int savings = d2 - d1 - cheat_cost;
                if (savings >= min_savings) {
                    ++count;
                }
            }
        }
    }

    return count;
}

long long part1(const Grid& grid, const Point& start, const Point& end) {
    DistMap dist = trace_path(grid, start, end);
    return count_cheats(dist, 2, 100);
}

long long part2(const Grid& grid, const Point& start, const Point& end) {
    DistMap dist = trace_path(grid, start, end);
    return count_cheats(dist, 20, 100);
}

int main() {
    std::ifstream file("../input.txt");
    if (!file) {
        std::cerr << "Error: Cannot open input file" << std::endl;
        return 1;
    }

    std::stringstream buffer;
    buffer << file.rdbuf();
    std::string input = buffer.str();

    auto [grid, start, end] = parse_grid(input);

    std::cout << "Part 1: " << part1(grid, start, end) << std::endl;
    std::cout << "Part 2: " << part2(grid, start, end) << std::endl;

    return 0;
}
